astacus/coordinator/plugins/cassandra/plugin.py

```diff
@@ -162,12 +151,36 @@ def get_restore_schema_from_snapshot_steps(self, *, context: OperationContext, r
             table_glob=SNAPSHOT_GLOB,
             keyspaces_to_skip=[ks for ks in SYSTEM_KEYSPACES if ks != "system_schema"],
             match_tables_by=ipc.CassandraTableMatching.cfid,
+            expect_empty_target=True,
+        )
+        restore_deltas_req = ipc.CassandraRestoreSSTablesRequest(
+            table_glob=BACKUP_GLOB,
+            keyspaces_to_skip=[ks for ks in SYSTEM_KEYSPACES if ks != "system_schema"],
+            match_tables_by=ipc.CassandraTableMatching.cfid,
+            expect_empty_target=False,
         )
 
         return [
             base.RestoreStep(storage_name=context.storage_name, partial_restore_nodes=req.partial_restore_nodes),
             CassandraRestoreSubOpStep(op=ipc.CassandraSubOp.restore_sstables, req=restore_sstables_req),
+            base.DeltaManifestsStep(json_storage=context.json_storage),
+            base.RestoreDeltasStep(
+                json_storage=context.json_storage,
+                storage_name=context.storage_name,
+                restore_delta_url=f"cassandra/{ipc.CassandraSubOp.restore_sstables}",
+                partial_restore_nodes=req.partial_restore_nodes,
+                restore_delta_request=restore_deltas_req,
+            ),
             restore_steps.StopReplacedNodesStep(partial_restore_nodes=req.partial_restore_nodes, cassandra_nodes=self.nodes),
+            restore_steps.UploadFinalDeltaStep(json_storage=context.json_storage, storage_name=context.storage_name),
+            base.RestoreDeltasStep(
```

Comment:
We should avoid directly reusing the same class multiple time in the steps list, that makes progress reporting more confusing. I'd rather see two named subclasses for `RestoreDeltasStep` that "configure" the base class, or just call non-step functions, and only pass truly dynamic parameters to these steps.

(This is also true for `CassandraRestoreSubOpStep`, and I don't think that should propagate further).